src/common/utxobased/keymanager/cleaners.ts

```diff
@@ -0,0 +1,173 @@
+ *
+ * (spec: https://github.com/EdgeApp/edge-core-js/blob/master/docs/key-formats.md)
+ */
+export interface PrivateKey {
+  coinType: number
+  format: CurrencyFormat
+  seed: string
+}
+export function asPrivateKey(
+  coinName: string,
+  coinType: number = 0
+): Cleaner<PrivateKey> {
+  return asCodec(
+    raw => {
+      if (raw == null || typeof raw !== 'object') {
+        throw new TypeError('Private keys must be objects')
+      }
+      const asCoinType = asOptional(asNumber, coinType)
+      return {
+        coinType: asCoinType(raw.coinType),
+        format: asOptionalPrivateKeyFormat(raw.format),
+        seed: asString(raw[`${coinName}Key`])
+      }
+    },
+    clean => {
+      const { coinType, format, seed } = clean
+      return {
+        coinType,
+        format,
+        [`${coinName}Key`]: seed
+      }
+    }
+  )
+}
+
+/**
+ * A cleaner for the public key format following the key-formats specification.
+ *
+ * (spec: https://github.com/EdgeApp/edge-core-js/blob/master/docs/key-formats.md)
+ */
+export interface PublicKey {
+  publicKeys: {
+    [format in CurrencyFormat]?: string
+  }
+}
+export const asPublicKey: Cleaner<PublicKey> = asObject({
+  publicKeys: asObject({
+    bip32: asOptional(asString),
+    bip44: asOptional(asString),
+    bip49: asOptional(asString),
+    bip84: asOptional(asString)
+  })
+})
+
+/**
+ * This utility returns the a wallet's supported formats as specified in the
+ * key-formats specification.
+ */
+export const getSupportedFormats = (
+  format: CurrencyFormat
+): CurrencyFormat[] => {
+  switch (format) {
+    case 'bip32':
+      return ['bip32']
+    case 'bip44':
+      return ['bip44']
+    case 'bip49':
+      return ['bip49', 'bip84']
+    case 'bip84':
+      return ['bip49', 'bip84']
+    default:
+      throw new Error(`Unsupported format ${format}`)
+  }
+}
+
+/**
+ * A cleaner that desensitizes the walletInfo object, excluding sensitive
+ * keys (seed/mnemonic, sync key, data key, etc). By using this object type
+ * internally within the plugin, we can minimize risk.
+ *
+ * It also includes internal derived data (publicKey, format, supportedFormats, etc).
+ * This derived data is to be used internally within the plugin and saved to disk (publicKey).
+ */
+export interface NumbWalletInfo {
+  id: string
+  type: string
+  keys: {
+    coinType: number
+    format: CurrencyFormat
+    supportedFormats: CurrencyFormat[]
+    publicKey: PublicKey
+  }
+}
+export const asNumbWalletInfo = (
+  pluginInfo: PluginInfo
+): Cleaner<NumbWalletInfo> => {
+  return (walletInfo: any) => {
+    const { engineInfo, coinInfo } = pluginInfo
+    const asCoinType = asOptional(asNumber, coinInfo.coinType)
+
+    // Prepare fields to return for NumbWalletInfo:
+    const { id, type } = walletInfo
+    const coinType = asCoinType(walletInfo.keys.coinType)
+    const format = asOptionalCurrencyFormat(walletInfo.keys.format)
+    const supportedFormats = getSupportedFormats(format)
```

Comment:
I didn't notice this before, but this logic won't work in watch-only mode. It assumes we have a private key, which isn't a given.

I think this function could be split into two sections like:

```js
const private = asMaybe(asCurrencyPrivateKey)(walletInfo.keys)
if (private != null) {
  // convert private to public
  return { format: private.format, ... }
}

const public = asMaybe(asCurrencyPublicKey)(walletInfo.keys)
if (public != null) {
  const formats = Object.keys(public.publicKeys).filter(asWalletFormat)
  return { format: formats[0], supportedFormats: formats, ... }
}

// If we get here, nothing matched:
throw new Error('either a public or private key are required')
```

This structure gets rid of the IIFE, and since we just try the keys both ways an go with whichever works.

This would allow `asNumbWalletInfo` to work in watch-only mode. Right now it will just blow up.